tests/acceptance/standard/test_schedules.py

```diff
@@ -199,6 +199,95 @@ def test__delete_schedule(self, project, schedules):
         )
         assert schedule is None
 
+    def test__schedule_enforce_new_schedule(self, project, schedules):
+        new_schedule = f"""
+        projects_and_groups:
+          {project.path_with_namespace}:
+            schedules:
+              enforce: true
+              "New schedule to test enforce config":
+                ref: main
+                cron: "30 1 * * *"
+        """
+
+        schedules_before = project.pipelineschedules.list()
+
+        run_gitlabform(new_schedule, project)
+
+        schedules_after = project.pipelineschedules.list()
+
+        schedule = self.__find_pipeline_schedule_by_description_and_get_first(
+            project, "New schedule to test enforce config"
+        )
+        assert len(schedules_before) == 6
+        assert len(schedules_after) == 1
+        assert schedule is not None
+        assert schedule.description == "New schedule to test enforce config"
+        assert schedule.ref == "main"
+        assert schedule.cron == "30 1 * * *"
+        assert schedule.cron_timezone == "UTC"
+        assert schedule.active is True
+
+    def test__schedule_enforce_new_and_existing_schedule(self, project, schedules):
+        new_schedule = f"""
+        projects_and_groups:
+          {project.path_with_namespace}:
+            schedules:
+              enforce: true
+              "New schedule to test enforce config":
+                ref: main
+                cron: "30 1 * * *"
+              "New schedule with variables":
+                ref: main
+                cron: "30 1 * * *"
+                variables:
+                    var1:
+                        value: value123
+                    var2:
+                        value: value987
+                        variable_type: file
+        """
+
+        schedules_before = project.pipelineschedules.list()
+
+        run_gitlabform(new_schedule, project)
+
+        schedules_after = project.pipelineschedules.list()
+
+        schedule1 = self.__find_pipeline_schedule_by_description_and_get_first(
+            project, "New schedule to test enforce config"
+        )
+        schedule2 = self.__find_pipeline_schedule_by_description_and_get_first(
+            project, "New schedule with variables"
+        )
+        assert len(schedules_before) == 1
+        assert len(schedules_after) == 2
```

Comment:
@gdubicki - It seems the tests are currently setup to be dependent on earlier test case's results. That's why this assertion looks odd to me. I expected each test case to be self-contained and not affecting other tests. I'm not sure if this was done intentionally. Thought I'd mention.

Reply:
This is a result of the fact that we are often using fixtures that have a pytest scope higher than "function". Particularly, [the fixture for a project has scope "class"](https://github.com/gitlabform/gitlabform/blob/main/tests/acceptance/conftest.py#L102-L103), so the schedules created in one method of a test class such as TestSchedules will in fact be retained in the next test method.

To fix that we would have to use "function" scope for an almost perfect separation of each test but we are are talking about tests on a real-world GitLab instance where creating and deleting projects in GitLab is pretty slow.

I just did a test of how this would work out in practice for `test_schedules.py`:

Code as-is, with project and group having scope "class":

1st run - 7.708s
2nd run - 6.280s
3rd run - 7.390s

Code with project and group scope changed to "function":

1st run - 14.556s
2nd run - 18.286s
3rd run - 18.335s

So the tests time effectively increases 2-2.5 times.

The total current time of running tests in GitHub Actions is ~3 minutes for standard and ~1.5 minutes for premium ones, so including the ~4 minutes of GitLab provisioning time the whole pipeline time would increase from ~7 minutes to ~10 minutes...

Because of that we are now using 2 ways of dealing with the issue of tests affecting each other, only where it affects the tests:

a) use "function" scoped fixtures on specific single tests - see the uses of `project_for_function` fixture,
b) make the assertions about the initial state more relaxed - see https://github.com/gitlabform/gitlabform/blob/main/tests/acceptance/standard/test_members.py#L12-L12 as an example.

Of course this is not perfect, but it's the best compromise between correctness and performance of the tests that I was able to develop at the time. If you think that we should make changes then feel free to share your ideas, @amimas. :)

Reply:
Thanks @gdubicki for the details and the performance tests. I also realized this came down to fixture scope after looking into it a bit, but didn't have time to learn what changes are needed if the scope is changed.

I'm starting to think what you've got right now actually makes sense. These are not mocked unit tests. As you said, we're running these against a real GitLab instance. Also not sure if there's a real need for running only individual test as opposed to running all tests in a single file. That would probably make the testing and development faster, but is it worth it? So, it is a balance between isolated tests vs performance of entire test suites.

I don't have any other ideas at the moment. Being new to the code base, it was a bit confusing at the beginning. I had to go through existing test cases and then figure out if the assertions in the new test case makes sense. Maybe what might help others is if we have additional sections in the contributing docs about general setup/expectations of acceptance tests. I have only played with 1 test file through this PR so far. If I can think of any improvements to the docs/development guide in future, I'll open new PR accordingly.

Reply:
I agree, we should at least document it. I'll try to do it soon.